---
title: "GNSC3301 Project"
author: "allie derry"
date: "April 14, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

## So far in this project, we have done research on how to get genetic code from NCBI and put it into R Studio, how to put the genome sequence into a string of characters, and find start and stop codons for that sequence. Here is a link that I have been basing my research off of: https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter7.html -Allison Derry

## Like Allie said, most of the current work has been researching ways to create our own form of a biopython capable of recognizing DNA and associate it with amino acids. Currently we are working on creating an amino acid table that will define codons as specific amino acids. I have been using the data carpentry resource to learn how biopython works and hopefully work backwards from there. (https://datacarpentry.org/semester-biology/materials/biopython-data-types-and-methods/) -Lauren Harris

## Since the HGP completed in 2003, widespread genetic data has been avaliable. R makes configuring this data efficient by using multiple programs that the article "Integrated analysis of genetic data with R" displays. With futher research into the programs, I believe I will be able to find the information needed to correctly code the data. Biopythos is included in this list of programs and the aricle "Bioipython: freely available Python tools for computational molecular biology and bioinformatics" shows many promising theroies to attempt. 
Biopython-https://academic.oup.com/bioinformatics/article/25/11/1422/330687 
R-https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3525150/
Emma Downing
